atc/exec/set_pipeline_step.go

```diff
@@ -130,7 +130,40 @@ func (step *SetPipelineStep) run(ctx context.Context, state RunState) error {
 		return nil
 	}
 
-	team := step.teamFactory.GetByID(step.metadata.TeamID)
+	var team db.Team
+	if step.plan.Team == "" {
+		team = step.teamFactory.GetByID(step.metadata.TeamID)
+	} else {
+		fmt.Fprintln(stderr, "\x1b[1;33mWARNING: specifying the team in a set_pipeline step is experimental and subject to change!\x1b[0m")
+		fmt.Fprintln(stderr, "")
+		fmt.Fprintln(stderr, "\x1b[33mcontribute to discussion #5731 with feedback: https://github.com/concourse/concourse/discussions/5731\x1b[0m")
+		fmt.Fprintln(stderr, "")
+
+		currentTeam, found, err := step.teamFactory.FindTeam(step.metadata.TeamName)
+		if err != nil {
+			return err
+		}
+		if !found {
+			return fmt.Errorf("team %s not found", step.metadata.TeamName)
+		}
+
+		targetTeam, found, err := step.teamFactory.FindTeam(step.plan.Team)
+		if err != nil {
+			return err
+		}
+		if !found {
+			return fmt.Errorf("team %s not found", step.plan.Team)
+		}
+
+		if !currentTeam.Admin() {
```

Comment:
Only allowing admin user to set pipeline to other team doesn't help much for our clusters. Can we do this? If target team and current team have shared owner/member, then allow to set.

Reply:
Would you be able to confirm what you mean by "shared owner/member"?

I interpret this in two ways:
1. There is one or more shared owners in `team-a` and `team-b`
2. The users and groups of owners in `team-a` and `team-b` are the same

I think (1) is not a good idea, because then anyone owner in `team-a` can set pipelines in `team-b` without necessarily having access to do it themselves.

I think (2) is probably what you meant, so the resultant logic would be 

```
if !currentTeam.Admin() && !currentTeam.Auth().OwnersAreSameAs(targetTeam.Auth()) {
  // error not permitted
}

// continue to set pipeline
```

where `atc.TeamAuth.OwnersAreSameAs` is a new function which would check that the users and groups are the same, disregarding ordering.

---

I'm happy to update my change to include this, though first I'll wait for @vito to weigh in with his opinion

Reply:
@tlwr I meant 1. My theory is that, say user-1 and user-2 are owner/member (member role has permission to set pipeline) of team-a; and user-2 and user-3 are owner/member of team-b. Then user-2 can get a pipeline from team-a and set to team-b, though user-1 cannot, but that still implies that team-a can set pipeline to team-b. I agree that opens a hole where user-1 can set-pipeline to team-b via a `set-pipeline` step from team-a.

For 2, I feel that "exact same" hurts usability. For example, team-a has owner user-1, and team-b has owner user-1 and user-2, in this case, I think team-a should still be able to set pipeline to team-b.

So combining 1 and 2, can we do this: if owners/members of team-a is a sub-set of team-b's owners/members, then team-a can set pipeline to team-b. Which is a little bit more flexible than 2.

@vito A real use case, I don't want to mention name of the project here. The project has ~100 concourse teams, one team is pipeline maintainer, rest of teams reflect to modules of the project. Because every module has similar CI/CD workflows, the pipeline maintainer team maintains a set of pipeline templates via vars, and rolls out pipelines to other teams using templates+vars. So the owners of the maintainer team are owners of other teams, but they are not admin users. Now they use the concourse-fly resource type, this feature will make their maintainer pipeline a little bit simpler.

Reply:
@evanchaoli I get you now, I can see how that would work:

`team-a` can set `team-b`'s pipelines if every owner/member of `team-a` is an owner/member in `team-b`

I can update this PR to include this behaviour, but I'll let you and @vito find agreement first :)

Reply:
@evanchaoli Ah, got it, thanks! I think we have a similar scenario between our team and another team which maintains open-source license scanning pipelines for a bunch of other teams.

Since we don't want `team:` to be here forever, I'd like to plan ahead and see how this use case might be supported in a future without it just to make sure we don't paint ourselves into a corner.

Let's call the team that maintains the pipelines for other teams "upstream" and the teams that consume the pipelines "downstream". Given that the goal is for the upstream team to be able to configure pipelines for the downstream teams, I think this could be modeled as a Project that all the downstream teams configure within their own Project.

So the downstream team's Project config would look something like this:

```yaml
plan:
- set_pipeline: foo
- set_pipeline: bar
- set_project: upstream-baz
```

I'm glossing over a lot of things here, but architecturally this seems sound: it allows the downstream teams to opt-in to the upstream project, and allows the upstream project to self-update when changes are made without requiring any further maintenance from the downstream teams, and without requiring special access to it. In fact they could live on completely separate clusters.

Do you think this kind of flow would make sense for your use case?

Reply:
@vito and @evanchaoli the above discussed behaviour is a superset of what is currently implemented in this PR, so that this doesn't hang around I'm happy to raise a follow up PR implementing whatever behaviour meets the use case

I've documented the behaviour currently implemented in this PR here https://github.com/concourse/docs/pull/341

Reply:
@tlwr Sounds fair! This can be resolved in a later PR.